### PR TITLE
Static Firewall Deployments

### PR DESCRIPTION
**Pull request by vknabel**
## Description

Goal:

- validate CRDs for firewalls
- static only
- focus on the contract between the metalstackcluster and the metalstackfirewalldeployment controller
- provide the structure for the future integration with the firewall-controller-manager
- keep stuff that will be thrown away simple (like the firewall creation)

Non-Goal:

- rolling of firewalls
- handling any edge case that the fcm will handle

References:

- https://github.com/metal-stack/firewall-controller-manager/pull/77

TODO:

- [x] At least some basic controller tests

When creating a cluster with the default templates, you will receive the following additional resources:

```
NAMESPACE                           NAME                                                                         CLUSTER         TEMPLATE        DEPLOYMENT                                               READY   AGE
e2e-integration-testing-2-basic-0   metalstackfirewalldeployment.infrastructure.cluster.x-k8s.io/basic-cluster   basic-cluster   basic-cluster   metal://partition1/00000000-0000-0000-0000-7cc25554232d   true    94s

NAMESPACE                           NAME                                                                       PARTITION   PROJECT                                IMAGE                 SIZE            AGE
e2e-integration-testing-2-basic-0   metalstackfirewalltemplate.infrastructure.cluster.x-k8s.io/basic-cluster   partition1   5535aee2-5f30-49e1-b636-93bc47373179   firewall-ubuntu-3.0   n3-medium-x86   94s
```

Has been integration tested (using local manifests):

```bash
$ make generate manifests test-e2e E2E_LABEL_FILTER="teardown,basic && \!move" IMG_TAG=feature-simplified-firewalls
...
Ran 3 of 8 Specs in 529.766 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 5 Skipped
PASS

Ginkgo ran 1 suite in 8m51.396969265s
Test Suite Passed
```

### Noteworthy

```NOTEWORTHY
Starting with this release the cluster-api-provider-metal-stack supports the automatic deployment of static firewalls. All default templates are now adjusted accordingly.
You can try this out by using the new `MetalStackFirewallTemplate` and `MetalStackFirewallDeployment` resources and reference the deployment within `MetalStackCluster` using the new `spec.firewallDeploymentRef.name` field.
```

### CODE REVIEW

File: config/manager/kustomization.yaml [hunk 1]
```diff
@@ -4,5 +4,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: controller
-  newName: ghcr.io/metal-stack/capms-controller
+  newName: ghcr.io/metal-stack/cluster-api-metal-stack-controller
```
[the remark this answers is not in the document]
Comment (vknabel) [reply config/manager/kustomization.yaml]: Should be better with #96 , this should be correct here.

File: internal/controller/metalstackfirewalldeployment_controller.go [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+			return nil
+		}
+
+		fwdeployList := &v1alpha1.MetalStackFirewallDeploymentList{}
+		err := r.Client.List(ctx, fwdeployList, &client.ListOptions{
+			Namespace: fwtemp.Namespace,
+			FieldSelector: fields.SelectorFromSet(fields.Set{
+				"spec.firewallTemplateRef.name": fwtemp.Name,
+			}),
+		})
```
[the remark this answers is not in the document]
Comment (vknabel) [reply internal/controller/metalstackfirewalldeployment_controller.go]: The firewall templates might not have the owner reference yet as they are added on first use (as the capi core provider does too)

File: internal/controller/metalstackfirewalldeployment_controller.go [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+	var err error
+
+	if !controllerutil.ContainsFinalizer(r.firewallDeployment, v1alpha1.FirewallDeploymentFinalizer) {
+		r.log.Info("finalizer not present, skipping deletion flow")
+		return nil
+	}
+
+	r.log.Info("reconciling resource deletion flow")
+
+	err = r.ensureAllMetalStackMachinesAreGone()
```
[the remark this answers is not in the document]
Comment (vknabel) [reply internal/controller/metalstackfirewalldeployment_controller.go]: Yes, cluster api does some cleanup and requires the control to be reachable.

File: config/clusterctl-templates/cluster-template-calico.yaml [hunk 4]
```diff
@@ -33,6 +33,68 @@ spec:
...
+          - 53
+          - 123
+        protocol: UDP
+        to:
+          - 0.0.0.0/0
+
+    ingress:
+      - comment: allow incoming HTTP and HTTPS traffic
+        ports:
+          - 80
```
[the remark this answers is not in the document]
Comment (vknabel) [reply config/clusterctl-templates/cluster-template-calico.yaml]: This is not only for the kube-apiserver but also for some basic ingresses that are likely to be deployed.

File: internal/controller/metalstackfirewalldeployment_controller.go [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+			Ingress: ingressRules,
+		},
+	}), nil)
+	if err != nil {
+		return fmt.Errorf("error creating firewall deployment: %w", err)
+	}
+
+	r.log.Info("created firewall deployment", "firewallID", fwresp.Payload.ID)
+
+	r.firewallDeployment.Spec.ManagedResourceRef = &v1alpha1.MetalStackManagedResourceRef{
```
Comment (Gerrit91) [inline internal/controller/metalstackfirewalldeployment_controller.go]: This approach could lead to leaking resources when storing the managed resource ref goes wrong (because there is no lookup if the firewall was already created). I guess it's just temporary anyway so we can address it later when we replace it with the firewall deployment?

Comment (vknabel) [reply internal/controller/metalstackfirewalldeployment_controller.go]: Jup, this is just temporary. Later we will lookup the firewall deployment resource.

File: internal/controller/metalstackfirewalldeployment_controller.go [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+func (r *MetalStackFirewallDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := mgr.GetCache().IndexField(context.TODO(), &v1alpha1.MetalStackFirewallDeployment{}, "spec.firewallTemplateRef.name", func(obj client.Object) []string {
+		fwdeploy, ok := obj.(*v1alpha1.MetalStackFirewallDeployment)
+		if !ok {
+			return nil
+		}
+		return []string{fwdeploy.Spec.FirewallTemplateRef.Name}
+	})
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (vknabel) [reply internal/controller/metalstackfirewalldeployment_controller.go]: True, it exists anyway as this error could not be recovered.
